manticore/core/workspace.py

```diff
@@ -0,0 +1,458 @@
+    '''
+    def __init__(self, desc=None):
+        '''
+        Create an object capable of producing Manticore output.
+
+        :param desc: A descriptor ('type:uri') of where to write output.
+        '''
+        self._store = _create_store(desc)
+        self._last_id = 0
+
+    @property
+    def uri(self):
+        return self._store.uri
+
+    def save_testcase(self, state, message=''):
+        '''
+        Save the environment from `state` to storage. Return a state id
+        describing it, which should be an int or a string.
+
+        :param State state: The state to serialize
+        :param int testcase_id: Identifier for the state
+        :param str state: Optional message to include
+        :return: A state id representing the saved state
+        '''
+
+        self.save_summary(state, message)
+        self.save_trace(state)
+        self.save_constraints(state)
+        self.save_input_symbols(state)
+        self.save_syscall_trace(state)
+        self.save_fds(state)
+        key = 'test_{:08x}.pkl'.format(self._last_id)
+        self._store.save_state(state, key)
+
+        self._last_id += 1
+
+    def save_stream(self, *rest):
+        return self._store.save_stream(*rest)
+
+    @contextmanager
+    def _named_stream(self, name):
+        '''
+        Create an indexed output stream i.e. '00000001_name'
+
+        :param name: Identifier for the stream
+        :return: A context-managed stream-like object
+        '''
+
+
+        with self._store.save_stream('test_{:08x}.{}'.format(self._last_id, name)) as s:
+            yield s
+
+    def save_summary(self, state, message):
+        with self._named_stream('messages') as summary:
+            summary.write("Command line:\n  '{}'\n" .format(' '.join(sys.argv)))
+            summary.write('Status:\n  {}\n\n'.format(message))
+
+            memories = set()
+            for cpu in filter(None, state.platform.procs):
+                idx = state.platform.procs.index(cpu)
+                summary.write("================ PROC: %02d ================\n"%idx)
+                summary.write("Memory:\n")
+                if hash(cpu.memory) not in memories:
+                    summary.write(str(cpu.memory).replace('\n', '\n  '))
+                    memories.add(hash(cpu.memory))
+
+                summary.write("CPU:\n{}".format(cpu))
+
+                if hasattr(cpu, "instruction") and cpu.instruction is not None:
+                    i = cpu.instruction
+                    summary.write("  Instruction: 0x%x\t(%s %s)\n"%(
+                        i.address, i.mnemonic, i.op_str))
+                else:
+                    summary.write("  Instruction: {symbolic}\n")
+
+    def save_trace(self, state):
+        with self._named_stream('trace') as f:
+            for pc in state.context['visited']:
+                f.write('0x{:08x}\n'.format(pc))
+
+    def save_constraints(self, state):
+        # XXX(yan): We want to conditionally enable this check
```

Comment:
We may also just remove it(assert) as it's going to fail later when trying to solve the paths constraint anyway